File: courier-service/src/main/java/com/gokhana/courierservice/bus/publisher/KafkaMessageProducerManager.java
package com.gokhana.courierservice.bus.publisher;

import com.gokhana.courierservice.model.event.CourierEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.util.concurrent.ListenableFuture;
import org.springframework.util.concurrent.ListenableFutureCallback;

@Service
public class KafkaMessageProducerManager implements MessageProducerService {

    private static final Logger LOG = LoggerFactory.getLogger(KafkaMessageProducerManager.class);

    private final KafkaTemplate<String, Object> kafkaTemplate;
    @Value(value = "${kafka.topic.courierEvent}")
    private String courierEventTopic;

    public KafkaMessageProducerManager(KafkaTemplate<String, Object> kafkaTemplate) {
        this.kafkaTemplate = kafkaTemplate;
    }

    @Override
    @Async
    public void sendCourierEvent(CourierEvent courierEvent) {
        LOG.debug("courierEvent object is sending.. -> {}", courierEvent);

        ListenableFuture<SendResult<String, Object>> future = kafkaTemplate.send(courierEventTopic, courierEvent);
        future.addCallback(new ListenableFutureCallback<>() {

            @Override
            public void onSuccess(SendResult<String, Object> result) {
                LOG.debug("Sent message='{}' with offset={}", courierEventTopic, result.getRecordMetadata().offset());
            }

            @Override
            public void onFailure(Throwable ex) {
                LOG.error("Unable to send message='{}'", courierEventTopic, ex);
            }

        });
    }

}

File: order-service/src/main/kotlin/com/gokhana/orderservice/bus/publisher/KafkaMessageProducerManager.kt
package com.gokhana.orderservice.bus.publisher

import com.gokhana.orderservice.model.event.OrderEvent
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.support.SendResult
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.util.concurrent.ListenableFutureCallback

@Service
class KafkaMessageProducerManager(val kafkaTemplate: KafkaTemplate<String, Any>) : MessageProducerService {

    private val LOG = LoggerFactory.getLogger(KafkaMessageProducerManager::class.java)


    @Value(value = "\${kafka.topic.orderEvent}")
    private val orderEventTopic: String? = null


    @Async
    override fun sendOrderEventMessage(orderEvent: OrderEvent?) {
        LOG.debug("orderEvent object is sending.. -> $orderEvent")
        val future = kafkaTemplate.send(
            orderEventTopic!!, orderEvent
        )
        future.addCallback(object : ListenableFutureCallback<SendResult<String?, Any?>?> {
            override fun onSuccess(result: SendResult<String?, Any?>?) {
                LOG.debug(
                    "Sent message='{}' with offset={}",
                    orderEventTopic,
                    result!!.recordMetadata.offset()
                )
            }

            override fun onFailure(ex: Throwable) {
                LOG.error("Unable to send message='{}'", orderEventTopic, ex)
            }
        })
    }

}